enterprise/mqtt/src/main/java/io/crate/mqtt/protocol/MqttProcessor.java

```diff
@@ -0,0 +1,170 @@
+import io.netty.handler.codec.mqtt.MqttPublishMessage;
+import io.netty.handler.codec.mqtt.MqttQoS;
+import io.netty.handler.codec.mqtt.MqttVersion;
+import io.netty.handler.timeout.IdleStateHandler;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+
+import java.util.Locale;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+
+
+public class MqttProcessor {
+
+    private static final Logger LOGGER = Loggers.getLogger(MqttProcessor.class);
+
+
+    private final MqttIngestService ingestService;
+
+    public MqttProcessor(MqttIngestService ingestService) {
+        this.ingestService = ingestService;
+    }
+
+    public void handleConnect(Channel channel, MqttConnectMessage msg) {
+        String clientId = msg.payload().clientIdentifier();
+        LOGGER.info("CONNECT for client <{}>", clientId);
+
+        if (!(msg.variableHeader().version() == MqttVersion.MQTT_3_1.protocolLevel()
+                || msg.variableHeader().version() == MqttVersion.MQTT_3_1_1.protocolLevel())) {
+            sendErrorResponse(channel, MqttConnectReturnCode.CONNECTION_REFUSED_UNACCEPTABLE_PROTOCOL_VERSION)
+                    .addListener(f -> LOGGER.warn("CONNECT sent UNACCEPTABLE_PROTOCOL_VERSION"));
+            return;
+        }
+
+        /*
+         If you don’t need a state to be hold by the broker,
+         in MQTT 3.1.1 (current standard) it is also possible to send an empty ClientId,
+         which results in a connection without any state.
+         A condition is that clean session is true, otherwise the connection will be rejected.
+         http://www.hivemq.com/blog/mqtt-essentials-part-3-client-broker-connection-establishment
+         */
+        if (clientId == null || clientId.length() == 0) {
+            if (!msg.variableHeader().isCleanSession()) {
+                sendErrorResponse(channel, MqttConnectReturnCode.CONNECTION_REFUSED_IDENTIFIER_REJECTED)
+                        .addListener(cf -> LOGGER.warn("CONNECT sent IDENTIFIER_REJECTED"));
+                return;
+            }
+            clientId = UUID.randomUUID().toString();
+            LOGGER.info("Client connected with server generated identifier: {}", clientId);
+        }
+        NettyUtils.clientID(channel, clientId);
+
+        int keepAlive = msg.variableHeader().keepAliveTimeSeconds();
+        if (keepAlive > 0) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Client connected with keepAlive of {} s", keepAlive);
+            }
+            NettyUtils.keepAlive(channel, keepAlive);
+
+            if (channel.pipeline().names().contains("idleStateHandler")) {
+                channel.pipeline().remove("idleStateHandler");
+            }
+            channel.pipeline().addFirst("idleStateHandler", new IdleStateHandler(0, 0, Math.round(keepAlive * 1.5f)));
```

Comment:
Why * 1.5?

Reply:
Avoid declaring the channel idle premature possibly. Do you have a suggestion for a better factor?

Reply:
No, but a short comment along the lines of `Give some slack to avoid terminating connections too early` would be nice.